### DISCUSSION TITLE
multiple "FQDN vs IP" tuples in /etc/ipsec.secrets bring to PSK overwrite

### DISCUSSION THREAD

**Discussion by marcogaravello** [strongswan/strongswan#995] [Q&A]
Hello,

I recently experienced this kind of issue: if in _/etc/ipsec.secrets_ I set this

```
#vpn_tunnel_for_customer_1
myvpn.mydomain.com peer_IP_of_customer_1 : PSK password1
peer_IP_of_customer_1 myvpn.mydomain.com : PSK password1

#vpn_tunnel_for_customer_2
myvpn.mydomain.com peer_IP_of_customer_2 : PSK password2
peer_IP_of_customer_2 myvpn.mydomain.com : PSK password2
```

vpn_tunnel_for_customer_1 WON'T work, while vpn_tunnel_for_customer_2 DOES.
Why?
Because on vpn_tunnel_for_customer_1, ipsec is using password2 instead of password1 (I tested this).
This seems due to the fact I'm using FQDN "myvpn.mydomain.com" in the configuration, instead of my public IP. In fact, if I switch using my public IP, issue disappears.
It looks like, using FQDN, only one password can be stored in memory and so the last one overwrites the previous ones.
Is this behaviour known?

Thanks in advance.

**Comment by tobiasbrunner** [comment]
Please post the config and logs (see #196). Note that the order of identities assigned to a secret doesn't matter, FQDN identities are not resolved, and a single identity match is enough to consider a secret. So if you configure your local identity, which is shared between multiple connections, for every secret, all will match. However, if you configure your local IP that is later not used as identity, that's completely irrelevant and basically as if you hadn't configured it (as long as the remote identity matches).  As responder, all matching secrets will be tried. What secret will be used as initiator depends on how well the remote identity matches (if it's known at all yet).

**Reply by marcogaravello** [reply]
mmm sorry just to clarify If I get your point: in your opinion, cfg of PSKs I reported on my first point should work with no issue and with no overlap?
Or do you already see something wrong/incorrect on it?

**Reply by tobiasbrunner** [reply]
They are redundant and it's usually better to not configure the local identity. Other than that it really depends on your config etc.

**Reply by marcogaravello** [reply]
Dear,

ok understood
We used to configure the local identity (and redundant too) to overcome to eventual troubles with non-strongswan customer peers (some of them doesn't start tunnel if local identity is not specified, and sometimes they need it redundant, too).

Anyway, I'm attaching my /etc/ipsec.conf with relevant sections.
[ipsec.conf_cut.txt](https://github.com/strongswan/strongswan/files/8511441/ipsec.conf_cut.txt)

**Reply by tobiasbrunner** [reply]
> and sometimes they need it redundant, too

Not sure what you mean, because, as I said, configuring them twice is useless as the order of assigned identities doesn't matter.

> Anyway, I'm attaching my /etc/ipsec.conf with relevant sections.

Since you configured `rightid=%any`, no remote identity is available as initiator (no IDr will be sent for IKEv2 either). That means only secrets that match the local identity will be considered. However, if you configure multiple with the same local identity, all of them. You should definitely consider configuring `rightid` to the expected identity of the peer (and then use that identity for the secrets), or at least use different local identities for the connections to select the correct secret. Also note that one of the connections uses IKEv1 while the other uses IKEv2. You can't directly compare the two. In particular, there is a fallback lookup for IKEv1 that uses the IP addresses. That's not the case for IKEv2, which only uses identities.

**Reply by marcogaravello** [reply]
Okay I got the point. So could it be that

> if you configure multiple with the same local identity, all of them

in my corner case, initiator does NOT consider all of them while processing IKEv1 tunnel cfg, but only the last one?

**Reply by tobiasbrunner** [reply]
Sorry, not sure what you mean exactly.

**Reply by marcogaravello** [reply]
I'll try to better explain myself: as local identity is reported more than once, you wrote that all related PSKs are considered.
Said so, could it be that initiator does NOT consider all of them while establishing IKEv1 tunnel, but initiator wrongly consider only the last PSK (the one associated to local identity)?

**Reply by tobiasbrunner** [reply]
The initiator can only select a single PSK, so yes it only considers one of them (it prefers better matching PSKs based on local and remote identities, but if there is no remote identity, it basically comes down to the order in which they were loaded/stored).

**Reply by marcogaravello** [reply]
Ok that is clear. So issue is mainly due to the fact that I set rightid=%any, so if I set rightid for each conn I should get rid of it, forcing initiator to a sigle perfect match.
But with rightid=%any in conn %default, if rightid is not specified in the "real" conn sections, which value assumes rightid for each connection?

Thanks so much.

**Reply by tobiasbrunner** [reply]
> But with rightid=%any in conn %default, if rightid is not specified in the "real" conn sections, which value assumes rightid for each connection?

Whatever the peer sends will be its identity, no specific identity is enforced.

**Reply by marcogaravello** [reply]
ah ok, instead if I set rightid=customer_1_name, initiator will not care about identity sent by customer peer, but use customer_1_name as identity and looks for it in /etc/ipsec.secrets, correct?

**Reply by tobiasbrunner** [reply]
No, it will enforce that identity. If the peer responds with anything but `customer_1_name`, the authentication will fail. So you better configure the actual identity of the peer.

**Reply by marcogaravello** [reply]
aah ok.
How can I obtain/log identity sent by customer peer, and the PSK selected by my initiator for each connection attempt?

Thanks again

**Reply by tobiasbrunner** [reply]
The identity you see in the log, however, only if the correct PSK was selected in the first place (ask the person responsible for the other config for the identity to avoid any trial and error). The PSK will simply be the one whose associated identities match the configured local and remote identities best.

**Reply by marcogaravello** [reply]
Thanks
Can you make me an example of where identity sent by right (customer) peer is logged?

**Reply by marcogaravello** [reply]
Hello,

in the meantime I did what you suggested: I specified rightid parameter for both VPN tunnels ([ipsec.conf_cut_with_rightids.txt](https://github.com/strongswan/strongswan/files/8590355/ipsec.conf_cut_with_rightids.txt)), but issue is still there: for both tunnels it seems that PSK selected by initiator is password2

> #vpn_tunnel_for_customer_1
> myvpn.mydomain.com peer_IP_of_customer_1 : PSK password1
> peer_IP_of_customer_1 myvpn.mydomain.com : PSK password1
> 
> #vpn_tunnel_for_customer_2
> myvpn.mydomain.com peer_IP_of_customer_2 : PSK password2
> peer_IP_of_customer_2 myvpn.mydomain.com : PSK password2

Did I still misunderstand something?
Thanks in advance

**Reply by tobiasbrunner** [reply]
While those PSKs are still redundant (the order of identities doesn't matter), as is configuring the local identity if it's the same for every PSK, there shouldn't be a reason for the second PSK to get selected for the first config. What strongSwan version do you use? You might also want to make sure there are not typos and check the log.

**Reply by marcogaravello** [reply]
ipsec --version returns

`Linux strongSwan U5.8.2/K5.13.0-1017-aws`

In the log I get not so related errors but I suppose they depend on how the remote host manages wrong key event e.g. in this case the peer does not answer and forces initiator to send retrasmissions:

> Apr 29 10:59:10 05[CFG] received stroke: initiate 'vpn_tunnel_for_customer_1'
> Apr 29 10:59:10 14[IKE] <vpn_tunnel_for_customer_1|350> initiating Main Mode IKE_SA vpn_tunnel_for_customer_1[350] to peer_IP_of_customer_1
> Apr 29 10:59:10 14[ENC] <vpn_tunnel_for_customer_1|350> generating ID_PROT request 0 [ SA V V V V V ]
> Apr 29 10:59:10 14[NET] <vpn_tunnel_for_customer_1|350> sending packet: from 10.160.0.239[500] to peer_IP_of_customer_1[500] (180 bytes)
> Apr 29 10:59:10 12[NET] <vpn_tunnel_for_customer_1|350> received packet: from peer_IP_of_customer_1[500] to 10.160.0.239[500] (188 bytes)
> Apr 29 10:59:10 12[ENC] <vpn_tunnel_for_customer_1|350> parsed ID_PROT response 0 [ SA V V V V V ]
> Apr 29 10:59:10 12[IKE] <vpn_tunnel_for_customer_1|350> received NAT-T (RFC 3947) vendor ID
> Apr 29 10:59:10 12[IKE] <vpn_tunnel_for_customer_1|350> received DPD vendor ID
> Apr 29 10:59:10 12[ENC] <vpn_tunnel_for_customer_1|350> received unknown vendor ID: 82:99:03:17:57:a3:60:82:c6:a6:21:de:00:00:00:00
> Apr 29 10:59:10 12[IKE] <vpn_tunnel_for_customer_1|350> received FRAGMENTATION vendor ID
> Apr 29 10:59:10 12[IKE] <vpn_tunnel_for_customer_1|350> received FRAGMENTATION vendor ID
> Apr 29 10:59:10 12[CFG] <vpn_tunnel_for_customer_1|350> selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1024
> Apr 29 10:59:10 12[ENC] <vpn_tunnel_for_customer_1|350> generating ID_PROT request 0 [ KE No NAT-D NAT-D ]
> Apr 29 10:59:10 12[NET] <vpn_tunnel_for_customer_1|350> sending packet: from 10.160.0.239[500] to peer_IP_of_customer_1[500] (268 bytes)
> Apr 29 10:59:10 15[NET] <vpn_tunnel_for_customer_1|350> received packet: from peer_IP_of_customer_1[500] to 10.160.0.239[500] (252 bytes)
> Apr 29 10:59:10 15[ENC] <vpn_tunnel_for_customer_1|350> parsed ID_PROT response 0 [ KE No NAT-D NAT-D ]
> Apr 29 10:59:10 15[IKE] <vpn_tunnel_for_customer_1|350> local host is behind NAT, sending keep alives
> Apr 29 10:59:10 15[ENC] <vpn_tunnel_for_customer_1|350> generating ID_PROT request 0 [ ID HASH ]
> Apr 29 10:59:10 15[NET] <vpn_tunnel_for_customer_1|350> sending packet: from 10.160.0.239[4500] to peer_IP_of_customer_1[4500] (92 bytes)
> Apr 29 10:59:14 09[IKE] <vpn_tunnel_for_customer_1|350> sending retransmit 1 of request message ID 0, seq 3
> Apr 29 10:59:14 09[NET] <vpn_tunnel_for_customer_1|350> sending packet: from 10.160.0.239[4500] to peer_IP_of_customer_1[4500] (92 bytes)
> Apr 29 10:59:21 09[IKE] <vpn_tunnel_for_customer_1|350> sending retransmit 2 of request message ID 0, seq 3
> Apr 29 10:59:21 09[NET] <vpn_tunnel_for_customer_1|350> sending packet: from 10.160.0.239[4500] to peer_IP_of_customer_1[4500] (92 bytes)



P.S.: ok they are redundant, I'll remember that and eventually delete one row each as local identity is always the same.
Thank you!

**Reply by tobiasbrunner** [reply]
More interesting would be the log messages when the daemon starts, where the config and secrets are loaded. But since you replace IPs, we still can't be sure there are no typos, you alone can check for that (or just remove those local identities so PSKs only match if the remote IP/identity actually matches).

That this message can't be sent is normal if the PSK is incorrect as the responder won't be able to decrypt it. However, that could also be because it has the wrong PSK configured/selected.

**Reply by marcogaravello** [reply]
Dear,

thanks
I checked and there are not typos
I also made this test: I modified "password2" and from that moment both tunnel cannot be established then.
Which test do you suggest next?

Regarding removing local identity I'm a bit worried because I cannot be sure of the remote identity set on customer side. But if you make me an example of what you mean exactly with "remove those local identities" I can try anyway.

**Reply by tobiasbrunner** [reply]
> Which test do you suggest next?

Does the IKEv1 connection work if you remove/comment the second PSK (i.e. if only the first PSK is defined)?

> because I cannot be sure of the remote identity set on customer side

Then this will not work anyway. If you configure `rightid` it **has** to match the peer's identity (it will default to the IP address if you don't define it).

> But if you make me an example of what you mean exactly with "remove those local identities" I can try anyway.

Simply `peer_IP_of_customer_1 : PSK password1` (or replace `peer_IP_of_customer_1` with whatever identity the peer uses, it does not have to be an IP address).

**Reply by marcogaravello** [reply]
Dear Tobias,

first of all, sorry for the long waiting!
In the end, I work this out testing your suggestions:

> Does the IKEv1 connection work if you remove/comment the second PSK (i.e. if only the first PSK is defined)?

Yes, it works anyway.
Is this strange also for you? Or did you expect that?

Then:

> Then this will not work anyway. If you configure rightid it has to match the peer's identity (it will default to the IP address if you don't define it).

So if in _ipsec.conf_ I set `rightid` for a conn, that must match peer identity set on right (customer) side, correct?
What if it doesn't match? Connection is refused?
Anyway: is there a way to retrieve/log the value of the right peer identity?

In the end:

> Simply peer_IP_of_customer_1 : PSK password1 (or replace peer_IP_of_customer_1 with whatever identity the peer uses, it does not have to be an IP address).

I tried and IKEv1 tunnel works (I think because there's the fallback lookup on IP, as you explained before), while IKEv2 fails, because right peer identity doesn't equal to IP I suppose..

Thanks in advance!

**Reply by tobiasbrunner** [reply]
>> Does the IKEv1 connection work if you remove/comment the second PSK (i.e. if only the first PSK is defined)?
> Yes, it works anyway.
> Is this strange also for you? Or did you expect that?

Yes, that's fine because it confirms that the first PSK is the correct one for that connection. The weird thing was that the second PSK was apparently selected when there were two (you previously reported that changing the second PSK broke both connections). Of course, with the local ID assigned to both, either one is eligible, but why it would prefer the second over the first is weird. But again there might be some mismatch with the IPs that we didn't see.

> So if in ipsec.conf I set rightid for a conn, that must match peer identity set on right (customer) side, correct?
> What if it doesn't match? Connection is refused?

Yes, either the sent remote identity doesn't match on the peer and it will reject the connection attempt with e.g. NO_PROPOSAL_CHOSEN. Or the connection is aborted locally later on when the identity returned by the peer that doesn't match the configuration.

> I tried and IKEv1 tunnel works (I think because there's the fallback lookup on IP, as you explained before), while IKEv2 fails, because right peer identity doesn't equal to IP I suppose..

Yes, sounds reasonable (unless the identity for IKEv1 is the IP anyway). For IKEv2, you should configure the identity you set in `rightid` of that connection (this works for IKEv1 too, as initiator, not as responder because the identity won't be known until the packet is decrypted using the right PSK).

**Reply by marcogaravello** [reply]
> Of course, with the local ID assigned to both, either one is eligible, but why it would prefer the second over the first is weird. But again there might be some mismatch with the IPs that we didn't see.

This is still an open point for me, but now not so important because thanks to your help I specified rightid parameter for each tunnel cfg in ipsec.conf and now issue is gone: basically I specified rightid filling it with corresponing peer_IP_of_customer_X value, and with some tests I verified that now correct PSK is selected.

> Yes, either the sent remote identity doesn't match on the peer and it will reject the connection attempt with e.g. NO_PROPOSAL_CHOSEN. Or the connection is aborted locally later on when the identity returned by the peer that doesn't match the configuration.

Mmm ok clear, but is there a way to retrieve/log the value of the right peer identity?

> Yes, sounds reasonable (unless the identity for IKEv1 is the IP anyway).

Sorry not clear to me what you mean for "unless the identity for IKEv1 is the IP anyway"

**Reply by tobiasbrunner** [reply]
> This is still an open point for me,

We'd need a minimal reproducer config (ipsec.conf and ipsec.secrets) to replicate and debug this problem. If you could provide one, that'd be great.

> Mmm ok clear, but is there a way to retrieve/log the value of the right peer identity?

Only if the connection progresses this far. If the peer rejects the connection attempt, you won't see it (configuring `%any` as remote identity will avoid sending one with IKEv2, if the peer accepts that, it might respond with its actual identity, which will then be logged). IKEv1 should log the received identity too (e.g. if no matching config is found).

> Sorry not clear to me what you mean for "unless the identity for IKEv1 is the IP anyway"

I meant that there is no extra fallback lookup etc. if the IKEv1 identity is the IP address anyway. Then it will just be a straight identity match.

**Reply by marcogaravello** [reply]
> We'd need a minimal reproducer config (ipsec.conf and ipsec.secrets) to replicate and debug this problem. If you could provide one, that'd be great.

Here you are:
[ipsec.conf_cut.txt](https://github.com/strongswan/strongswan/files/8928886/ipsec.conf_cut.txt)
[ipsec.secrets.txt](https://github.com/strongswan/strongswan/files/8928916/ipsec.secrets.txt)
I just realized that in the past (https://github.com/strongswan/strongswan/discussions/995#discussioncomment-2660797) I tried following your same advice - specifying rightid for each tunnel - but (for a mistake of mine I suppose) situation didn't improve, while as you read in my previous post I did exactly the same and issue is gone!
Maybe first time I forgot "ipsec reread something" and/or "ipsec reload", d'oh.
Thanks even more for your time debugging this anyway!

> Only if the connection progresses this far. If the peer rejects the connection attempt, you won't see it (configuring %any as remote identity will avoid sending one with IKEv2, if the peer accepts that, it might respond with its actual identity, which will then be logged). IKEv1 should log the received identity too (e.g. if no matching config is found).

Thanks, nice tip
Have you got a IKEv1 and IKEv2 log cut where this info appear? So I can look for same pattern in my log and identify it.

**Reply by tobiasbrunner** [reply]
Thanks. OK, the reason this config doesn't work is `rightid=%any`. If that's removed, the IKEv1 connection works even without explicit `rightid` for each connection (the IKEv2 connection works either way).

There are two aspects that affect this:
1. Shared secrets loaded later are actually preferred. I guess, the idea is that "newer" credentials with the same identities indirectly replace "older" ones. That's more useful for other credentials e.g. certificates and CRLs but was also changed for shared secrets (they can be loaded individually e.g. via _vici_ plugin). I changed this over 11 years ago (75827463fa9a4961f9c151e09da2cb4329dd38d1), so I guess it generally didn't cause issues due to the identity matching (I didn't even remember that the order is like that).
2. If `rightid` is not specified, the identity is replaced with the remote IP before the PSK is looked up. Actually, it happens very early, when the legacy _stroke_ plugin loads the config and the remote identity is just set to whatever is configured in `right` (could be a FQDN or an IP address or even an invalid value if there is more than one IP address/subnet configured). So in this case both identities match the first two PSKs perfectly and the connection works. However, if `rightid` is explicitly set to `%any`, it's not replaced when the config is loaded and the lookup matches all PSKs with the same preference (as all specify the local identity and the remote identity matches for none). But due to the order, the last PSK in the file is matched first, so that's what's selected and the connection fails.

**Reply by marcogaravello** [reply]
Interesting.
So summarizing (please be patient, I could understand wrong!):

for both IKEs (v1 and v2):
- if `rightid `is not specified at all, remote identity is replaced with value of `right `parameter and match is performed looking for `leftid` and `right` value pair
- if `rightid` is set to `%any`, no replacement is made so match is based only on `leftid` parameter
- if `rightid` is specified, match is performed looking for `leftid` and `rightid` value pair

NOTE: last loaded matching PSK always wins.
NOTE2: using IKEv1, identities have to be firstly decrypted so during first connection initiator IP is used to match PSK on responder side.

**Reply by tobiasbrunner** [reply]
The first part is correct, but I want to point out that the first bullet point really only applies to ipsec.conf-based configs. With swanctl.conf, there is no such replacement if no remote identity is configured.

NOTE is only correct for PSKs with the same preference based on the identities. PSKs that match better are still preferred.

NOTE2 is only correct if you mean that as an instruction to users in case the remote identity is not known. Because otherwise there is no difference between the first and any following connection attempts if the same config is used.

**Reply by marcogaravello** [reply]
All clear, thank you.

Regarding NOTE2:  to be sure of remote identity, have you got a IKEv1 and IKEv2 log cut where this info appear? So I can look for same pattern in my log and identify it.

**Reply by tobiasbrunner** [reply]
Check the [testresults](https://www.strongswan.org/testing/testresults/) for sample logs (mostly successful ones, though).

**Reply by marcogaravello** [reply]
Very interesting
but on which category and test do you expect remote identity appears to be logged?

Thanks so much in advance
